homeassistant/components/systemmonitor/sensor.py

```diff
@@ -158,183 +181,289 @@ def check_required_arg(value):
 ]
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+@dataclass
+class SensorData:
+    """Data for a sensor."""
+
+    argument: Any
+    state: str | None
+    value: Any | None
+    update_time: datetime.datetime | None
+    last_exception: BaseException | None
+
+
+async def async_setup_platform(
+    hass: HomeAssistant,
+    config: ConfigType,
+    async_add_entities: Callable,
+    discovery_info: Any | None = None,
+) -> None:
     """Set up the system monitor sensors."""
-    dev = []
+    entities = []
+    sensor_registry: dict[str, SensorData] = {}
+
     for resource in config[CONF_RESOURCES]:
+        type_ = resource[CONF_TYPE]
         # Initialize the sensor argument if none was provided.
         # For disk monitoring default to "/" (root) to prevent runtime errors, if argument was not specified.
         if CONF_ARG not in resource:
-            resource[CONF_ARG] = ""
+            argument = ""
             if resource[CONF_TYPE].startswith("disk_"):
-                resource[CONF_ARG] = "/"
+                argument = "/"
+        else:
+            argument = resource[CONF_ARG]
 
         # Verify if we can retrieve CPU / processor temperatures.
         # If not, do not create the entity and add a warning to the log
         if (
-            resource[CONF_TYPE] == "processor_temperature"
-            and SystemMonitorSensor.read_cpu_temperature() is None
+            type_ == "processor_temperature"
+            and await hass.async_add_executor_job(_read_cpu_temperature) is None
         ):
             _LOGGER.warning("Cannot read CPU / processor temperature information")
             continue
 
-        dev.append(SystemMonitorSensor(resource[CONF_TYPE], resource[CONF_ARG]))
+        sensor_registry[type_] = SensorData(argument, None, None, None, None)
+        entities.append(SystemMonitorSensor(sensor_registry, type_, argument))
+
+    scan_interval = config.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+    await async_setup_sensor_registry_updates(hass, sensor_registry, scan_interval)
+
+    async_add_entities(entities)
+
+
+async def async_setup_sensor_registry_updates(
```

Comment:
This looks very much like a data update coordinator.

Reply:
I started with that and it was a bit more complexity so I ended up pulling it back out.

Reply:
Fine to keep it like this.